src/App.tsx

```diff
@@ -1,26 +1,160 @@
+import { ErrorNotification } from './components/ErrorNotification';
+import {
+  getTodos,
+  addTodo,
+  USER_ID,
+  deleteTodo,
+  updateTodo,
+} from './api/todos';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [status, setStatus] = useState<FilterType>(FilterType.All);
+  const [errorMessage, setErrorMessage] = useState<Errors>(Errors.Default);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodo, setLoadingTodo] = useState<number[]>([]);
+
+  const filteredTodos = todos?.filter(todo => {
+    switch (status) {
+      case FilterType.Completed:
+        return todo.completed;
+      case FilterType.Active:
+        return !todo.completed;
+      default:
+        return true;
+    }
+  });
+
+  const unCompletedTodos = useMemo(
+    () => todos.filter(todo => !todo.completed).length,
+    [todos],
+  );
+
+  const completedTodos = useMemo(
+    () => todos.filter(todo => todo.completed).length,
+    [todos],
+  );
+
+  const allTodosCompleted = useMemo(
+    () => todos.every(todo => todo.completed),
+    [todos],
+  );
+
+  const onAddTodo = async (todoTitle: string) => {
+    setTempTodo({ id: 0, title: todoTitle, completed: false, userId: USER_ID });
+    try {
+      const newTodo = await addTodo({ title: todoTitle, completed: false });
+
+      setTodos(prev => [...prev, newTodo]);
+    } catch (err) {
+      setErrorMessage(Errors.AddTodo);
+      throw err;
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const onRemoveTodo = async (todoId: number) => {
+    setLoadingTodo(prev => [...prev, todoId]);
+    try {
+      await deleteTodo(todoId);
+      setTodos(prev => prev.filter(todo => todo.id !== todoId));
+    } catch (err) {
+      setErrorMessage(Errors.DeleteTodo);
+    } finally {
+      setLoadingTodo(prev => prev.filter(id => id !== todoId));
+    }
+  };
+
+  const onClearCompleted = async () => {
+    const completedTodo = todos.filter(todo => todo.completed);
+
+    completedTodo.forEach(todo => onRemoveTodo(todo.id));
+  };
+
+  const onUpdateTodo = async (todoToUpdate: Todo) => {
+    setLoadingTodo(prev => [...prev, todoToUpdate.id]);
+    try {
+      const updatedTodo = await updateTodo(todoToUpdate);
+
+      setTodos(prev =>
+        prev.map(todo => (todo.id === updatedTodo.id ? updatedTodo : todo)),
+      );
+    } catch (err) {
+      setErrorMessage(Errors.UpdateTodo);
+      throw err;
+    } finally {
+      setLoadingTodo(prev => prev.filter(id => id !== todoToUpdate.id));
+    }
+  };
+
+  const onSelectAll = async () => {
+    if (unCompletedTodos > 0) {
+      const activeTodos = todos.filter(todo => !todo.completed);
+
+      activeTodos.forEach(todo => {
+        onUpdateTodo({ ...todo, completed: true });
+      });
+    } else {
+      todos.forEach(todo => {
+        onUpdateTodo({ ...todo, completed: false });
+      });
+    }
+  };
```

Comment:
do we need `useCallback` in these cases?in which?

Reply:
So I can use _useCallback_ in all these variables?